pkg/raft/fsm/fsm.go

```diff
@@ -0,0 +1,102 @@
+package fsm
+
+import (
+	"bufio"
+	"encoding/base64"
+	"io"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/hashicorp/raft"
+	"github.com/rs/zerolog/log"
+)
+
+type CommunityStateMachine struct {
+	Log       []string `json:"log"`
+	StateJSON []byte
+}
+
+func NewCommunityStateMachine() *CommunityStateMachine {
+	return &CommunityStateMachine{
+		Log:       make([]string, 0),
+		StateJSON: []byte("{}"),
+	}
+}
+
+// Apply log is invoked once a log entry is committed.
+// It returns a value which will be made available in the
+// ApplyFuture returned by Raft.Apply method if that
+// method was called on the same Raft node as the FSM.
+func (sm *CommunityStateMachine) Apply(entry *raft.Log) interface{} {
+	sm.Log = append(sm.Log, string(entry.Data))
+	log.Info().Msgf("applying state transition %v", entry)
+
+	// Base64 decode so that no weird JSON quoting interferes along the way
+	decoded, err := base64.StdEncoding.DecodeString(string(entry.Data))
+	if err != nil {
+		log.Error().Msgf("error decoding log entry data: %s")
+	}
+
+	// TODO patching should be validated by the leader beforehand so there
+	// is no chance of an error here
+	patch, err := jsonpatch.DecodePatch(decoded)
+	if err != nil {
+		log.Error().Msgf("error decoding JSON patch: %s", err)
+	}
+
+	modified, err := patch.Apply(sm.StateJSON)
+	if err != nil {
+		log.Error().Msgf("error applying JSON patch: %s", err)
+	}
+
+	sm.StateJSON = modified
+
+	return nil
+}
+
+// Snapshot is used to support log compaction. This call should
+// return an FSMSnapshot which can be used to save a point-in-time
+// snapshot of the FSM. Apply and Snapshot are not called in multiple
+// threads, but Apply will be called concurrently with Persist. This means
+// the FSM should be implemented in a fashion that allows for concurrent
```

Comment:
hmm can you elaborat on how FSM allows for concurrent updates while a snapshot is happening? We should try to avoid multiple threads touching the same data structures when possible, since there can easily be lots of bugs created that way

Reply:
all the calls to the FSM interface that modify this data are made by the raft library, so we need to trust that they do it in a safe way.

Reply:
ahh I see. this is a raft comment